plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -1612,6 +1624,7 @@ private void finishWrite(ConnectorSession session, IcebergTableHandle table, Col
             if (!fullyDeletedFiles.isEmpty()) {
                 DeleteFiles deleteFiles = transaction.newDelete();
                 fullyDeletedFiles.keySet().forEach(deleteFiles::deleteFile);
+                getExtraSummaryMetadata(session).forEach(deleteFiles::set);
```

Comment:
isn't it quite error-prone to need to remember about `getExtraSummaryMetadata(session).forEach(xxx::set)` whenever we do an operation on a Table.

is there a way to help remember?

Reply:
Yeah I can make a function that commits `PendingUpdate`.

Some of the `commit()` calls in this file might not be `SnapshotUpdate<T>`. However `SnapshotUpdate<T>` is `PendingUpdate<Snapshot>`, so a commit function could be made that does a type check and then sets `extraSummaryMetadata`.

In Spark, we use `WriteBuilder` whiich limits the number of locations we need to add this call to.

Let me try a few different things now that I have a more comprehensive set of unit tests =)

Reply:
We can potentially use a `MethodInterceptor` from Guice to call the extra summary metadata whenever `PendingUpdate<?>::commit` is called.

I'm looking into how to do the binding for that now (and if it's possible to use Guice this way).

Reply:
Oh I just realized maybe the `ConnectorTableHandle` would be a good place to put this. Without changing the Iceberg interface at least (which is also a possibility we can bring up).

Iceberg's `BaseTransaction` recommits all of the pending updates / changes (the ones that are being called with `::set` for the extra metadata here), so maybe we can take advantage of that and just apply it all again and commit again in one place.

I'll explore that unless you have any input (or think the aspect idea is worth a shot). Or if you think a change to the Iceberg API is merited, that's not necessarily out of the realm of possibilities :)

Reply:
> We can potentially use a `MethodInterceptor` from Guice to call the extra summary metadata whenever `PendingUpdate<?>::commit` is called.

We don't use aspect programming, this usually leads to code that's very hard to reason about.

> Iceberg's BaseTransaction recommits all of the pending updates / changes (the ones that are being called with ::set for the extra metadata here), so maybe we can take advantage of that and just apply it all again and commit again in one place.

Can the transaction object record info for the snapshot metadata?
(i think it cannot, but please confirm)

Reply:
test comment